add Docker and Kubernetes how-to instructions (#945)
btcpayserver/btcpayserver-doc: docs/ReverseProxyToTor.md

Now, visiting `mydomain.com` should show your BTCPay Server instance.

## Do all this in a Docker container

Ready made docker image 

  https://hub.docker.com/r/cloudgenius/socator

> Code behing the image

Reviewer: "Behind" typo

Author: Thanks for catching.

> Code behing the image

Reviewer 2: ```suggestion
Code behind the image
```

Author: Thanks.

  https://github.com/beacloudgenius/socator

```
### SocaTor = SOCAT + TOR
Based on https://github.com/Arno0x/Docker-Socator

> It uses socat to listen on a given TCP port (5000 in this example) and to redirect incoming traffic to a tor hidden service specified through environment variables. Itacts as a relay between the standard web and a hidden service on the tor network. You can optionally restrict the IP addresses that are allowed to connect to this service by specifying an `ALLOWED_RANGE` environment variable and using CIDR notation.

Reviewer: Forgot space "it acts"

Author: Thanks for catching.

> It uses socat to listen on a given TCP port (5000 in this example) and to redirect incoming traffic to a tor hidden service specified through environment variables. Itacts as a relay between the standard web and a hidden service on the tor network. You can optionally restrict the IP addresses that are allowed to connect to this service by specifying an `ALLOWED_RANGE` environment variable and using CIDR notation.

Reviewer 2: ```suggestion
It uses socat to listen on a given TCP port (5000 in this example) and to redirect incoming traffic to a Tor hidden service specified through environment variables. It acts as a relay between the standard web and a hidden service on the Tor network. You can optionally restrict the IP addresses that are allowed to connect to this service by specifying an `ALLOWED_RANGE` environment variable and using CIDR notation.
```

### Usage

Break free from cloud services providers limitations, secure and protect your bitcoin full node, connect that with a BTC Pay server, all behind TOR.
Selectively expose the BTC Pay Server payment gateway and API to clearnet using socat+tor running on the Internet.

--------------

##### build

    docker build -t cloudgenius/socator .

##### push

    docker push cloudgenius/socator

##### Start the image in background (*daemon mode*) with IP address restriction:

    docker run -d \
        -p 5000:5000 \
        -e "ALLOWED_RANGE=192.168.1.0/24" \
        -e "TOR_SITE=zqktlwiuavvvqqt4ybvgvi7tyo4hjl5xgfuvpdf6otjiycgwqbym2qad.onion" \
        -e "TOR_SITE_PORT=80" \
        --name socator \
        cloudgenius/socator

##### Start the image in foreground:

    docker run --rm -ti \
        -p 5000:5000 \
        -e "TOR_SITE=zqktlwiuavvvqqt4ybvgvi7tyo4hjl5xgfuvpdf6otjiycgwqbym2qad.onion" \
        -e "TOR_SITE_PORT=80" \
        --name socator \
        cloudgenius/socato

Now http://localhost:5000 should show you the tor hidden service you specified in the above command.

```

> ## Use that Docker container in a Kubernetes Cluster using these manifests

Reviewer: Does BTCPay Server run inside Kubernetes? Didn't know we support that?

Reviewer 2: Hehe, me neither :)

@lvnilesh Can you use the code fence notation for the code samples as we do in the rest of the docs?

Author: I find it best for privacy and security to run a full node, NBxplorer, and BTCpayserver all behind TOR and completely in my physical possession and full control.

However, I still want to expose the payment endpoint to the public clearnet without asking BTC payers to go through TOR. 

Enter socator container running in kubernetes.

In my example, I am running socator container in kubernetes in the cloud and everything else runs in my control and supervision.

I will examine the code fence notation next and update the PR.

Author: code fence notation updated to match the rest.

Now, visiting `mydomain.com` should show your BTCPay Server instance.

## Do all this in a Docker container

Ready made docker image 

  https://hub.docker.com/r/cloudgenius/socator

Code behind the image

  https://github.com/beacloudgenius/socator

### SocaTor = SOCAT + TOR
Based on https://github.com/Arno0x/Docker-Socator

It uses socat to listen on a given TCP port (5000 in this example) and to redirect incoming traffic to a Tor hidden service specified through environment variables. It acts as a relay between the standard web and a hidden service on the Tor network. You can optionally restrict the IP addresses that are allowed to connect to this service by specifying an `ALLOWED_RANGE` environment variable and using CIDR notation.

### Usage

> Break free from cloud services providers limitations, secure and protect your bitcoin full node, connect that with a BTC Pay server, all behind TOR.

Reviewer 3: ```suggestion
Break free from cloud services providers limitations, secure and protect your bitcoin full node, connect that with a BTCPay Server, all behind TOR.
```

> Selectively expose the BTC Pay Server payment gateway and API to clearnet using socat+tor running on the Internet.

Reviewer 3: ```suggestion
Selectively expose the BTCPay Server payment gateway and API to clearnet using socat+tor running on the Internet.
```

Author: Thanks,

> --------------

Reviewer 3: remove

Author: Sorry I did not understand this comment. Please help.

Reviewer 3: normally we're not using these dividers in the docs but overall it looks good :)

##### build

```sh
docker build -t cloudgenius/socator .
```

##### push

```sh
docker push cloudgenius/socator

> ``

Reviewer 3: missing ` to close formatting

Author: Thanks. Yikes. it appears I don't know how to write.

##### Start the image in background (*daemon mode*) with IP address restriction:

```sh
docker run -d \
        -p 5000:5000 \
        -e "ALLOWED_RANGE=192.168.1.0/24" \
        -e "TOR_SITE=zqktlwiuavvvqqt4ybvgvi7tyo4hjl5xgfuvpdf6otjiycgwqbym2qad.onion" \
        -e "TOR_SITE_PORT=80" \
        --name socator \
        cloudgenius/socator
```

##### Start the image in foreground:

```sh
docker run --rm -ti \
        -p 5000:5000 \
        -e "TOR_SITE=zqktlwiuavvvqqt4ybvgvi7tyo4hjl5xgfuvpdf6otjiycgwqbym2qad.onion" \
        -e "TOR_SITE_PORT=80" \
        --name socator \
        cloudgenius/socato
```

Now http://localhost:5000 should show you the tor hidden service you specified in the above command.

## Use that Docker container in a Kubernetes Cluster using these manifests

```yaml

> ---

Reviewer 3: remove dividers and make each one separate yaml code block

Author: Manifests separated.

Reviewer 3: just suggesting this to make it more beginner friendly :)

Now, visiting `mydomain.com` should show your BTCPay Server instance.

## Do all this in a Docker container

> Ready made docker image 
> 
>   https://hub.docker.com/r/cloudgenius/socator
> 
> Code behind the image
> 
>   https://github.com/beacloudgenius/socator

Reviewer 2: ```suggestion
Ready made [docker image](https://hub.docker.com/r/cloudgenius/socator) ([Code](https://github.com/beacloudgenius/socator))
```

### SocaTor = SOCAT + TOR

> Based on https://github.com/Arno0x/Docker-Socator

Reviewer 2: ```suggestion
Based on [Docker-Socator](https://github.com/Arno0x/Docker-Socator)
```

It uses socat to listen on a given TCP port (5000 in this example) and to redirect incoming traffic to a Tor hidden service specified through environment variables. It acts as a relay between the standard web and a hidden service on the Tor network. You can optionally restrict the IP addresses that are allowed to connect to this service by specifying an `ALLOWED_RANGE` environment variable and using CIDR notation.

Please note: 

This container does not have any nginx component because Kubernetes provides for it. 

### Usage

Break free from cloud services providers limitations, secure and protect your bitcoin full node, connect that with a BTC Pay server, all behind TOR.

> Selectively expose the BTC Pay Server payment gateway and API to clearnet using socat+tor running on the Internet.

Reviewer 2: ```suggestion
Selectively expose the BTCPay Server payment gateway and API to clearnet using socat+tor running on the Internet.
```

--------------

##### build

```sh
docker build -t cloudgenius/socator .
```

##### push

```sh
docker push cloudgenius/socator
```

##### Start the image in background (*daemon mode*) with IP address restriction:

```sh
docker run -d \
        -p 5000:5000 \
        -e "ALLOWED_RANGE=192.168.1.0/24" \
        -e "TOR_SITE=zqktlwiuavvvqqt4ybvgvi7tyo4hjl5xgfuvpdf6otjiycgwqbym2qad.onion" \
        -e "TOR_SITE_PORT=80" \
        --name socator \
        cloudgenius/socator
```

##### Start the image in foreground:

```sh
docker run --rm -ti \
        -p 5000:5000 \
        -e "TOR_SITE=zqktlwiuavvvqqt4ybvgvi7tyo4hjl5xgfuvpdf6otjiycgwqbym2qad.onion" \
        -e "TOR_SITE_PORT=80" \
        --name socator \
        cloudgenius/socator
```

Now http://localhost:5000 should show you the tor hidden service you specified in the above command.

> 
> 
> 
> 

Reviewer 2: ```suggestion

```

## Use that Docker container in a Kubernetes Cluster using these manifests

These manifest assumes a typical Kubernetes cluster that exposes internal services (like socator running internallly at port 5000) to the clearnet/public internet via Nginx Ingress https://github.com/kubernetes/ingress-nginx and provide automated Let's Encrypt TLS/SSL certificates via https://github.com/jetstack/cert-manager. 

Deployment manifest

```yaml
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: socator
spec:
  replicas: 1
  selector:
    matchLabels:
      role: socator
  template:
    metadata:
      labels:
        role: socator
    spec:
      containers:
      - image: cloudgenius/socator # code https://github.com/beacloudgenius/socator
        imagePullPolicy: IfNotPresent
        name: socator
        env:
          - name: TOR_SITE

>             value: "zqktlwiuavvvqqt4ybvgvi7tyo4hjl5xgfuvpdf6otjiycgwqbym2qad.onion" # BTCpayserver tor address => docker exec tor cat /var/lib/tor/app-btcpay-server/hostname

Reviewer 2: ```suggestion
            value: "zqktlwiuavvvqqt4ybvgvi7tyo4hjl5xgfuvpdf6otjiycgwqbym2qad.onion" # BTCPay Server Tor address => docker exec tor cat /var/lib/tor/app-btcpay-server/hostname
```

          - name: TOR_SITE_PORT
            value: "80"
```

Service manifest

```yaml
---
apiVersion: v1
kind: Service
metadata:
  name: socator
spec:
  ports:
    - name: http
      port: 5000
  selector:
    role: socator
```

Ingress manifest

```yaml
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: socator
  annotations:
    kubernetes.io/ingress.class: nginx
    cert-manager.io/cluster-issuer: letsencrypt-prod
spec:
  rules:
    - host: btcpayserver.mydomain.com
      http:
        paths:
          - backend:
              service:
                name: socator
                port:
                  number: 5000
            path: /
            pathType: Prefix
  tls:
    - hosts:
        - btcpayserver.mydomain.com
      secretName: socator-tls

```

## Resources

> * [Nginx reverse proxy to .onion site in TOR network](https://itgala.xyz/nginx-reverse-proxy-to-onion-site-in-tor-network/)

Reviewer 2: ```suggestion
* [nginx reverse proxy to .onion site in Tor network](https://itgala.xyz/nginx-reverse-proxy-to-onion-site-in-tor-network/)
```

Now, visiting `mydomain.com` should show your BTCPay Server instance.

## Do all this in a Docker container

> Ready made [docker image](https://hub.docker.com/r/cloudgenius/socator) ([Code](https://github.com/beacloudgenius/socator))
> 

Reviewer 2: ```suggestion
Ready made [Docker image](https://hub.docker.com/r/cloudgenius/socator) ([Code](https://github.com/beacloudgenius/socator))
```

### SocaTor = SOCAT + TOR
Based on [Docker-Socator](https://github.com/Arno0x/Docker-Socator)

It uses socat to listen on a given TCP port (5000 in this example) and to redirect incoming traffic to a Tor hidden service specified through environment variables. It acts as a relay between the standard web and a hidden service on the Tor network. You can optionally restrict the IP addresses that are allowed to connect to this service by specifying an `ALLOWED_RANGE` environment variable and using CIDR notation.

Please note: 

This container does not have any nginx component because Kubernetes provides for it. 

> 
> 

Reviewer 2: ```suggestion

```

Please note: 

This container does not have any nginx component because Kubernetes provides for it. 

### Usage

Break free from cloud services providers limitations, secure and protect your bitcoin full node, connect that with a BTC Pay server, all behind TOR.
Selectively expose the BTCPay Server payment gateway and API to clearnet using socat+tor running on the Internet.

--------------

##### build

```sh
docker build -t cloudgenius/socator .
```

##### push

```sh
docker push cloudgenius/socator
```

##### Start the image in background (*daemon mode*) with IP address restriction:

```sh
docker run -d \
        -p 5000:5000 \
        -e "ALLOWED_RANGE=192.168.1.0/24" \
        -e "TOR_SITE=zqktlwiuavvvqqt4ybvgvi7tyo4hjl5xgfuvpdf6otjiycgwqbym2qad.onion" \
        -e "TOR_SITE_PORT=80" \
        --name socator \
        cloudgenius/socator
```

##### Start the image in foreground:

```sh
docker run --rm -ti \
        -p 5000:5000 \
        -e "TOR_SITE=zqktlwiuavvvqqt4ybvgvi7tyo4hjl5xgfuvpdf6otjiycgwqbym2qad.onion" \
        -e "TOR_SITE_PORT=80" \
        --name socator \
        cloudgenius/socator
```

> Now http://localhost:5000 should show you the tor hidden service you specified in the above command.

Reviewer 2: ```suggestion
Now http://localhost:5000 should show you the tor hidden service you specified in the above command.

```

## Use that Docker container in a Kubernetes Cluster using these manifests

These manifest assumes a typical Kubernetes cluster that exposes internal services (like socator running internallly at port 5000) to the clearnet/public internet via Nginx Ingress https://github.com/kubernetes/ingress-nginx and provide automated Let's Encrypt TLS/SSL certificates via https://github.com/jetstack/cert-manager. 

> 
> 

Reviewer 2: ```suggestion

```
